Refactor options and glyph icons

SummerWish opened the pull request:
Hi,

I refactored the options part and glyph part of the repo.

- Created privly-application/shared/javascripts/options.js

  This file provides a uniform interface to read and write all Privly extension options. I renamed local storage names of some options to keep consistency with other options. It will automatically rename old name values to new names when user updates the extension.

  This file is intended to act as a basic library, uses in background scripts and privly applications.

  This file also provides background message interfaces to read and write values (when it is inserted as a background script), which means content script can use Chrome message to access the option interface. Developers no longer need to write a background script function to pass values between localStorage and content script. This file has done the task.

  When option values are updated, it will send broadcast messages to all background scripts and content scripts. Those scripts can subscribe to option change messages, to respond user operations, with a very low degree of coupling. As you can see, I removed most of the code from `modal_button.js` and `reading_process.js`.

  In addition, I make *injection on/off* (in modal buton) as a kind of option, which means it will be preserved when user closes the browser.

- Created privly-application/shared/javascripts/glyph.js

  Like options.js, this file is intended to act as a basic library. It relies on options.js and provides functionalities related to glyph icons, for example, generate new glyph and create glyph HTML. No more duplicate glyph codes everywhere.

- Namespace `Privly`

  I introduced a namespace `Privly` for those two libraries (`options.js` is in `Privly.Options`, `glyph.js` is in `Privly.Glyph`). I think we can step by step move more *basic libraries* into this namespace to increase the readbility of the code.

The changes passes all unit tests in `privly-application`. Integration tests are all passed. Unit tests in `privly-chrome` are not modified in this PR since they have already been marked as `to modify` :-)

PR of privly-application: https://github.com/privly/privly-applications/pull/250

smcgregor commented on javascripts/background_scripts/modal_button.js:
```diff
@@ -1,64 +1,36 @@
-   */
-  badgeText: "on",
-
+function updateBrowserAction(enableInjection) {
```
Needs function comment.

smcgregor commented on javascripts/background_scripts/reading_process.js:
```diff
@@ -140,55 +36,36 @@ var readingProcess = {
-    } else if( url.indexOf("privlyApp=PlainPost") > 0) {
+    } else if (url.indexOf("privlyApp=PlainPost") > 0) {
       path = "privly-applications/PlainPost/show.html?privlyOriginalURL=";
-    } else if( url.indexOf("https://priv.ly") === 0 ) {
```
Try to keep whitespace changes in their own pull requests. It makes it easier to focus attention where it is needed.

SummerWish replied:
Got :-)

smcgregor commented:
This is really forcing an issue that I have wanted to clean up and document. We have many scripting contexts that are messaging each other and there is not a central listing of all the messages that are sent between the contexts. I can either make this resource and review/merge this pull request this week, or I can work on the embedded posting form pull request. Which is more important to you at this point?

SummerWish commented:
You can work on this PR :-) I'm going to transit embed posting to seamless posting and try to adapt it as a baic layer under Message application by the way.

smcgregor commented on javascripts/background_scripts/modal_button.js:
```diff
@@ -1,64 +1,36 @@
-// Handles the receipt of Privly URLs for addition to the web page.
-// The request object should contain the privlyUrl.
-chrome.extension.onMessage.addListener(modalButton.modeChange);
+updateBrowserAction(Privly.Options.isInjectionEnabled());
```
nit: newline at the end of the file.

smcgregor commented on javascripts/background_scripts/posting_process.js:
```diff
@@ -242,17 +242,5 @@ chrome.runtime.onMessage.addListener(
 
-// Respond to the request sent from posting_button.js with the value from localStorage["Options:DissableButton"]
-chrome.runtime.onMessage.addListener(
-  function(request, sender, sendResponse) {
```
I think this broke the ability to turn off the posting button.

SummerWish replied:
hi, posting button is changed to directly listen to option changed events:
https://github.com/privly/privly-chrome/pull/124/files#diff-d38f979261233fa789d1599f4b6f4feaR124

SummerWish replied:
correct: it doesn't listen to option changed events. It make message calls directly to the options API to get the option value.

smcgregor commented:
I think the option in the options page for deactivating the posting button is not working. Once you look into that and the renaming issue, I think this is good to merge. I am concerned that the "Privly" namespace is confusing since privly.js namespaces "privly", however, I think it would be good to improve name spacing across the whole project, if only because it will mean we don't need to do the "run_each" hack for running unit tests.

smcgregor commented:
Forget the button issue, that was a problem with my setup :)

smcgregor commented:
I just pushed this branch to to my fork so I can share test coverage information (when the build completes).

SummerWish commented:
Code updated :-)

smcgregor commented:
Let's increase the test coverage before merging: https://coveralls.io/builds/2667018/source?filename=shared%2Fjavascripts%2Foptions.js

SummerWish commented:
BTW, should we create `Privly.Storage` (local storage), which is similiar to `ls.js` but only JSON encode & JSON decode for all operations? My idea is to preserve `ls.js` and introduce new `Privly.Storage`. `ls.js` is only used to read old values for compatibility with reading existing option values. For local storage calls outside options, we migrate them to `Privly.Storage`. For local storage calls inside options, we only use `ls.js` to read old values when upgrading to new names.

SummerWish commented:
And of course, new code should not use `ls.js`.

smcgregor commented:
That would be fine. Make Privly.Storage call the ls library until we fully deprecate the ls library. We will need to port the Xul/JetPack storage mechanisms to the new storage adapter and this would be an intermediate step.

smcgregor commented:
Also, can you open a pull request against [this documentation](https://github.com/privly/privly.github.io/pull/35) listing all the messages you want supported? We need to be more formal about what information is being shared/broadcast/etc since these can have serious security implications.

Once you document this and add code coverage, I can work to integrate these changes with the Xul extension.

SummerWish commented:
Just pushed new commits.
- `Privly.Options` renamed to `Privly.options`
- `Privly.Glyph` renamed to `Privly.glyph`
- implemented `Privly.storage`
- added more test cases for `Privly.options`


https://github.com/privly/privly-applications/pull/250

smcgregor commented:
I just pushed commits to my fork to see what happens with the [TravisCI](https://travis-ci.org/smcgregor/privly-applications/builds/65841137) tests and the coverage stats. I will be on the road for the next week so I will be reviewing only (likely not much coding output), but I think you should be safe assuming the functionality found in this pull request. One item I think you need to add is enough documentation to the header of the context messenger that @hitesh96db and @sammyshj will know how to use it, complete with examples. You should also put console.warn statements anywhere you are uncertain about the implementation for a non-chrome platform and leave a note pointing to the example of the implementation for Chrome.

SummerWish commented:
Thank you! I am adding more documentation for those libraries now.

smcgregor commented:
[Test coverage information](https://coveralls.io/builds/2755322).

smcgregor commented:
I think these changes were incorporated in subsequent pull requests. Feel free to reopen if this is not the case.
